test/unit/profile_validation_test.rb

```diff
@@ -136,4 +136,312 @@ def test_language_binding_validation
     after = check_memory
     assert_memory(before, after)
   end
+
+  def test_cardinality_check
+    sd = FHIR::StructureDefinition.new
+
+    element = FHIR::ElementDefinition.new('min' => 0, 'max' => '1', 'path' => "test1")
+    nodes = []
+    sd.errors = []
+    sd.verify_cardinality(element, nodes)
+    assert_empty(sd.errors)
+
+    nodes = ["one"]
+    sd.errors = []
+    sd.verify_cardinality(element, nodes)
+    assert_empty(sd.errors)
+
+    nodes = [1, 2, 3]
+    sd.errors = []    
+    sd.verify_cardinality(element, nodes)
+    assert_equal("test1 failed cardinality test (0..1) -- found 3", sd.errors[0])
+
+    element = FHIR::ElementDefinition.new('min' => 1, 'max' => '1', 'path' => "test2")
+    nodes = []
+    sd.errors = []
+    sd.verify_cardinality(element, nodes)
+    assert_equal("test2 failed cardinality test (1..1) -- found 0", sd.errors[0])
+
+    nodes = ["one"]
+    sd.errors = []
+    sd.verify_cardinality(element, nodes)
+    assert_empty(sd.errors)
+
+    nodes = [1, 2, 3]
+    sd.errors = []    
+    sd.verify_cardinality(element, nodes)
+    assert_equal("test2 failed cardinality test (1..1) -- found 3", sd.errors[0])
+
+    element = FHIR::ElementDefinition.new('min' => 2, 'max' => '*', 'path' => "test3")
+    nodes = []
+    sd.errors = []
+    sd.verify_cardinality(element, nodes)
+    assert_equal("test3 failed cardinality test (2..Infinity) -- found 0", sd.errors[0])
+
+    nodes = ["one"]
+    sd.errors = []
+    sd.verify_cardinality(element, nodes)
+    assert_equal("test3 failed cardinality test (2..Infinity) -- found 1", sd.errors[0])
+
+    nodes = [1, 2, 3]
+    sd.errors = []    
+    sd.verify_cardinality(element, nodes)
+    assert_empty(sd.errors)
+  end
+
+  def test_maximum_string_length_check
+    FHIR::StructureDefinition.send(:public, :verify_element) # make the function public so we can test it
```

Comment:
I agree these tests are leaning heavily toward testing implementation rather than interface. Refactoring the validation logic to be elsewhere makes sense, but to do it well would be a big effort.